### PR TITLE
chore: snowflake debug duplicate messages

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Add support for logging sample duplicate rows for Snowflake. Currently, the plan is to deploy only to mattermost for a few hours and get sample messages which will help in debugging for duplicates.
- We are storing the duplicate data from both the main table and the staging table in a temporary file. Since these duplicates are coming continuously, plan is to just get one of the sample files.

## Linear Ticket

- Resolves PIPE-266

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/integrations/snowflake/snowflake.go [hunk 1]
```diff
@@ -521,63 +509,133 @@ func (sf *Snowflake) joinColumnsWithFormatting(columns []string, format string)
...
-		WHERE
-		  	RECEIVED_AT > (SELECT DATEADD(day,-?,current_date())) AND
-		  	ID IN (
-				SELECT ID
-				FROM `+stagingTable+`
-		  	)
-        ORDER BY RECEIVED_AT ASC
-		LIMIT ?;`,
+	rows, err := db.QueryContext(ctx, `
+		WITH mainCTE AS (
```
Comment (fracasula) [inline warehouse/integrations/snowflake/snowflake.go:528]: What does `CTE` stand for?

File: warehouse/integrations/snowflake/snowflake.go [hunk 2]
```diff
@@ -521,63 +509,133 @@ func (sf *Snowflake) joinColumnsWithFormatting(columns []string, format string)
...
-	ctx context.Context,
-	db *sqlmw.DB,
-	mainTableName,
-	stagingTableName string,
-) ([]duplicateMessage, error) {
+func (sf *Snowflake) trackDuplicateMessageInfo(
+	ctx context.Context, db *sqlmw.DB, mainTableName,
+	stagingTableName, sortedColumns string,
+) error {
 	if !lo.Contains(sf.config.debugDuplicateWorkspaceIDs, sf.Warehouse.WorkspaceID) {
```
Comment (fracasula) [inline warehouse/integrations/snowflake/snowflake.go:516]: [question] Is `workspaceID` too broad? Perhaps `destinationID` would be better?

File: warehouse/integrations/snowflake/snowflake.go [hunk 3]
```diff
@@ -521,63 +509,133 @@ func (sf *Snowflake) joinColumnsWithFormatting(columns []string, format string)
...
+			  SELECT
+				ID
+			  FROM
+				`+stagingTable+`
+			)
+		  ORDER BY
+			RECEIVED_AT ASC
+		  LIMIT
+			?
+		), StagingCTE AS (
```
Comment (fracasula) [inline warehouse/integrations/snowflake/snowflake.go:552]: [optional] Do we really need `StagingCTE`?

* In `mainCTE` we're getting all the records from the last X days that are also present in the staging table.
* In `stagingCTE` we're reading from the staging table instead but we select only the records that are already set as duplicate in `mainCTE`.

I understand that in the `UNION` at the end you want the identifier table but what for? If those records are duplicated they are in both tables, right?

Just saying, maybe just save the `mainCTE` and then add the staging file in the dump unless you find this way to be more convenient.

**Review by fracasula — APPROVED** [review]

File: warehouse/integrations/snowflake/snowflake.go (hunk 3 above)
Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:552]: They will have the same `messageID`, not other columns, that's what we will be figuring out from the sample.

File: warehouse/integrations/snowflake/snowflake.go (hunk 1 above)
Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:528]: Common Table Expression (CTE)

File: warehouse/integrations/snowflake/snowflake.go (hunk 2 above)
Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:516]: Initially, it was added only for `mattermost` use case for their workspace. We can probably extend it as well to destinationID.

File: warehouse/integrations/snowflake/snowflake.go (hunk 3 above)
Comment (fracasula) [reply warehouse/integrations/snowflake/snowflake.go:552]: Yeah that's why I said we could include the staging file in the dump, it's data we have already. I get it though, looks more convenient this way and the additional scanning is as big as the data from the staging file so it shouldn't be huge :+1:

**Pull request closed without merging**